src/core_plugins/kibana/public/dashboard/panel/dashboard_panel.js

```diff
@@ -0,0 +1,111 @@
+import React from 'react';
+import ReactDOM from 'react-dom';
+import PropTypes from 'prop-types';
+import classNames from 'classnames';
+
+import { DashboardViewMode } from '../dashboard_view_mode';
+import { PanelHeader } from './panel_header';
+
+export class DashboardPanel extends React.Component {
+
+  constructor(props) {
+    super(props);
+    this.state = {};
+    this.embeddable = null;
+    this.embeddableHandler = null;
+    this.parentNode = null;
+  }
+
+  async componentWillMount() {
+    const { getEmbeddableHandler, panel, getContainerApi } = this.props;
+
+    this.containerApi = getContainerApi();
+    this.embeddableHandler = getEmbeddableHandler(panel.type);
+    const editUrl = await this.embeddableHandler.getEditPath(panel.id);
+    const title = await this.embeddableHandler.getTitleFor(panel.id);
+
+    this.setState({ editUrl, title });
+
+    this.destroyEmbeddable = await this.embeddableHandler.render(
+      this.panelElement,
+      panel,
+      this.containerApi);
+  }
+
+  isViewOnlyMode() {
+    return this.props.dashboardViewMode === DashboardViewMode.VIEW || this.props.isFullScreenMode;
+  }
+
+  getParentNode() {
+    if (!this.parentNode) {
+      this.parentNode = ReactDOM.findDOMNode(this).parentNode;
+    }
+    return this.parentNode;
+  }
+
+  toggleExpandedPanel = () => this.props.onToggleExpanded(this.props.panel.panelIndex);
+  deletePanel = () => {
+    this.props.onDeletePanel(this.props.panel.panelIndex);
+  };
+  onEditPanel = () => window.location = this.state.editUrl;
+
+  /**
+   * Setting the zIndex on onFocus and onBlur allows popups, like the panel menu, to appear above other panels in the
+   * grid.
+   */
+  onFocus = () => {
+    this.getParentNode().style.zIndex = 1;
```

Comment:
Unfortunately I think moving the node into this component would be tough/impossible to do.  If I get rid of the `div` that surrounds `DashboardPanel` and put the key attribute on that, react-grid-layout doesn't work.  I think react-grid-layout requires the div on the outside, with the key attribute on it, to work properly.

If we used redux this would be easier since I could update the focusedPanel on state, and the parent could listen to that, and set the appropriate child with the z-index.  Without redux it's not easy for children to communicate with their parents.

I do want to move to redux as a next step.  Perhaps we can punt on improving this until then?  I do agree it's hacky...

Reply:
I agree that this feels like it breaks one of the core contracts of a React component. Why does the `zIndex` have to be set dynamically here? Couldn't it be just part of the stylesheet?

Reply:
I tried using react-grid-item:focus state to set the z-index but it doesn't work. Even though I click the menu option inside a grid item, it doesn't actually grab focus.  If I go into the css debugger and manually make it have focus, it works fine.

If parent selectors in css were possible I could do it, but they aren't supported yet, and may never be.

Reply:
> Without redux it's not easy for children to communicate with their parents.

Without Redux I've usually injected a function that gets called with the new value. I think the docs goes into this in https://facebook.github.io/react/docs/lifting-state-up.html. That could be an option, at least.

Reply:
Oh good point @kjbekkelund, I updated the logic so the grid is now handling it via functions passed to the child.